cmake/linking.cmake

```
@@ -0,0 +1,66 @@
+# Compressing the debug sections cuts the size of ttnn.so by half for builds with debug info.
+add_link_options($<$<CXX_COMPILER_ID:Clang>:-Wl,--compress-debug-sections=zstd>)
+
+# Use mold by default if it is available.
+if(NOT DEFINED CMAKE_LINKER_TYPE)
+    find_program(MOLD ld.mold)
+    if(MOLD)
+        message(STATUS "Linker not specified. Using mold linker: ${MOLD}")
+        set(CMAKE_LINKER_TYPE MOLD)
+    endif()
+endif()
+
+set(TT_LTO_ENABLED OFF)
+
+if(TT_ENABLE_LTO)
+    cmake_policy(SET CMP0069 NEW)
+    set(CMAKE_POLICY_DEFAULT_CMP0069 NEW)
+    include(CheckIPOSupported)
+    check_ipo_supported(RESULT result OUTPUT output)
+
+    # LTO can be requested but not supported, so handle that case
+    if(result)
+        message(STATUS "LTO/IPO is supported. Enabling for Release/RelWithDebInfo builds.")
+
+        # Do it this way to play nicer with ninja multi-config builds
+        # This enables clang's thinLTO by default.
+        set(CMAKE_INTERPROCEDURAL_OPTIMIZATION_RELEASE ON)
+        set(CMAKE_INTERPROCEDURAL_OPTIMIZATION_RELWITHDEBINFO ON)
+
+        set(TT_LTO_ENABLED ON)
+
+        # Enable one-definition-rule (ODR) warnings.
+        # Only works when LTO is enabled.
+        # FIXME: Forcing odr to warn until current errors are addressed. (#21850)
```

I'd like us to start adopting [google-readability-todo](https://clang.llvm.org/extra/clang-tidy/checks/google/readability-todo.html).  Not sure if it checks both TODOs and FIXMEs.
```suggestion
        # TODO(21850): Forcing odr to warn until current errors are addressed.
```